### PR TITLE
Rework the way to parse --set flag values

### LINKED ISSUE #3355: Truncated SSH Key
**Is this a request for help?**:
---
Yes

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
---
Issue

**What version of acs-engine?**:
---
0.18.9

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:
SSH Public Key was passed as a cli argument, end of the string is truncated.

**What you expected to happen**:
SSH Key is generated correctly.

**How to reproduce it** (as minimally and precisely as possible):
*...build acs-engine image...*
docker run microsoft/asc-engine:0.18.9 generate --api-model ./config/dev.json --set servicePrincipalProfile.clientID=... --set servicePrincipalProfile.secret=... --set linuxProfile.ssh.publicKeys[0].keyData="...w== matt@degree9.io" --output-directory ./clusters/dev

**Anything else we need to know**:
The `== matt@degree9.io` portion is truncated from the generated api model.

### LINKED ISSUE #3369: Command line arguments not properly handled - equal signs truncated
**Is this a request for help?**: NO

**Is this an ISSUE or FEATURE REQUEST?** ISSUE

**What version of acs-engine?**: 0.17.1 => 0.18.9 (both specifically tested)

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
ACS-Engine 0.18.9 with Kubernetes

**What happened**:
Secrets were removed from the API file and passed from command line; extract example below:
```
      "servicePrincipalProfile": {
        "clientId": "",
        "secret": ""   
      }
    }
````
```
acs-engine generate --output-directory "./_output/rack8s" --set servicePrincipalProfile.clientId='123a4321-c6eb-4b61-9d6f-7db123e14a7a',servicePrincipalProfile.secret='#msRock5!t=' ./clusterBuild-ACS_config-cardk8s.json
```

When this was done, the servicePrincipalClientSecret value in the generated azuredeploy.parameters.json has the '=' truncated out of it, failing without error.  If the value has a leading '=', the following exception is thrown:
```
INFO[0000] Error returned by LoadContainerService: json: cannot unmarshal number into Go struct field ServicePrincipalProfile.secret of type string
FATA[0000] error loading API model in generateCmd: error parsing the api model: json: cannot unmarshal number into Go struct field ServicePrincipalProfile.secret of type string
```

If, on the other hand, the same values are placed in the API config JSON (with or without a leading '='), the correct value shows up in the ARM template parameters file.

**What you expected to happen**:
I expected the command line parameters and those used in the API JSON to behave in the same manner.

**How to reproduce it** (as minimally and precisely as possible):
Example API JSON:
```
{
    "apiVersion": "vlabs",
    "properties": {
      "orchestratorProfile": {
        "orchestratorType": "Kubernetes",
        "orchestratorRelease": "1.10",
          "kubernetesConfig": {
            "networkPlugin": "kubenet",
            "networkPolicy": "calico",
            "privateCluster": {
			  "enabled": true          
			},
            "enableRbac": true,
            "maxPods": 80,
            "addons": [
              {
                "name": "tiller",
                "enabled": false
              },
              {
                "name": "kubernetes-dashboard",
                "enabled": false
              }
            ]
          }
        },
      "masterProfile": {
        "count": 1,
        "dnsPrefix": "rack8s",
        "vmSize": "Standard_D2_v2",
        "osDiskSizeGB": 200,
        "vnetSubnetId": "/subscriptions/123412341234/resourceGroups/cardinalsolutionsgroup-network/providers/Microsoft.Network/virtualNetworks/k8s-vnet/subnets/k8s-subnet",
        "firstConsecutiveStaticIP": "10.1.0.100",
        "vnetCidr": "10.1.0.0/16"    
      },
      "agentPoolProfiles": [
        {
          "name": "linuxpool1",
          "count": 2,
          "vmSize": "Standard_D2_v2",
          "availabilityProfile": "AvailabilitySet",
          "osDiskSizeGB": 200,
          "vnetSubnetId": "/subscriptions/123412341234/resourceGroups/jbh-network/providers/Microsoft.Network/virtualNetworks/k8s-vnet/subnets/k8s-subnet"
        }
      ],
      "linuxProfile": {
        "adminUsername": "",
        "ssh": {
          "publicKeys": [
            {
              "keyData": ""
            }
          ]
        }
      },
      "servicePrincipalProfile": {
        "clientId": "",
        "secret": ""   
      }
    }
  }
```
Example command line input:
```
 acs-engine generate --output-directory "./_output/rack8s" --set servicePrincipalProfile.clientId='123a1238-c6eb-4b61-9d6f-7db6f1e14123',servicePrincipalProfile.secret='=MsR0ck5!t=',aadProfile.serverAppID='123a1238-c6eb-4b61-9d6f-7db6f1e14123',aadProfile.clientAppID='123a1238-c6eb-4b61-9d6f-7db6f1e14123',aadProfile.tenantID='123a1238-c6eb-4b61-9d6f-7db6f1e14123',linuxProfile.adminUsername='cardinal',linuxProfile.ssh.publicKeys[0].keyData='ssh-rsa AAAAB3N123412341234123412341234123412341213412341234rElF9/uV2/+S6itZ' ./clusterBuild-ACS_config-cardk8s.json 
```
When the above command is run, it bombs on unmarshalling (due to the leading '=').  If the leading '=' is removed, the resulting servicePrincipalClientSecret will be `MsR0ck5!t` not `MsR0ck5!t=` as expected.  If these values are removed from the command line and placed in the API JSON, both `=MsR0ck5!t=` and `MsR0ck5!t=` work as expected.

**Anything else we need to know**: This is relevant because service principal secrets, as generated by the Azure Portal, often have trailing '=' signs.

### PR DESCRIPTION
**Pull request by jcorioland**
**What this PR does / why we need it**:
This PR fixes issues in parsing --set flag values, when '=' or ',' was provided in the values. 
Uses char by char values inspection instead of splitting.
Also adds some unit tests for this feature.

**Which issue this PR fixes** : fixes #3369 and fixes #3355 

**If applicable**:
- [ ] documentation
- [x] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:
```release-note
- fix for --set flag issue not parsing correctly key/values when an '=' or ',' char was present in the value.
```

### CODE REVIEW

File: cmd/generate.go [hunk 1]
```diff
@@ -104,10 +104,11 @@ func (gc *generateCmd) validate(cmd *cobra.Command, args []string) error {
 
 func (gc *generateCmd) mergeAPIModel() error {
 	var err error
-
+	log.Debugln("mergeAPIModel")
```
Comment (jackfrancis) [inline cmd/generate.go]: Are these log statements meant to be a part of the final PR?

**Comment by CecileRobertMichon** [comment]
@jcorioland how does this PR relate to https://github.com/Azure/acs-engine/pull/3579?

File: cmd/generate.go (hunk 1 above)
Comment (jcorioland) [reply cmd/generate.go]: nope, they are not. just pushed a fix. thanks!

**Comment by jcorioland** [comment]
@CecileRobertMichon I think this PR should be merged before #3579 as I changed the way I was parsing the command line parameters and merging the values with the JSON file.

**Pull request merged**